c-deps/libroach/mvcc.h

```diff
@@ -0,0 +1,636 @@
+      saved_buf_.resize(0);
+      saved_buf_.reserve(cur_raw_key_.size() + cur_value_.size());
+      saved_buf_.append(cur_raw_key_.data(), cur_raw_key_.size());
+      saved_buf_.append(cur_value_.data(), cur_value_.size());
+      cur_raw_key_ = rocksdb::Slice(saved_buf_.data(), cur_raw_key_.size());
+      cur_value_ = rocksdb::Slice(saved_buf_.data() + cur_raw_key_.size(), cur_value_.size());
+      rocksdb::Slice dummy_timestamp;
+      if (!SplitKey(cur_raw_key_, &cur_key_, &dummy_timestamp)) {
+        return setStatus(FmtStatus("failed to split mvcc key"));
+      }
+
+      // With the current iterator state saved we can move the
+      // iterator to the previous entry.
+      iter_rep_->Prev();
+      if (!iter_rep_->Valid()) {
+        // Peeking at the previous key should never leave the iterator
+        // invalid. Instead, we seek back to the first key and set the
+        // peeked_key to the empty key. Note that this prevents using
+        // reverse scan to scan to the empty key.
+        peeked_ = false;
+        *peeked_key = rocksdb::Slice();
+        iter_rep_->SeekToFirst();
+        return updateCurrent();
+      }
+    }
+
+    rocksdb::Slice dummy_timestamp;
+    if (!SplitKey(iter_rep_->key(), peeked_key, &dummy_timestamp)) {
+      return setStatus(FmtStatus("failed to split mvcc key"));
+    }
+    return true;
+  }
+
+  // clearPeeked clears the peeked flag. This should be called before
+  // any iterator movement operations on iter_rep_.
+  void clearPeeked() {
+    if (reverse) {
+      peeked_ = false;
+    }
+  }
+
+ public:
+  DBIterator* const iter_;
+  rocksdb::Iterator* const iter_rep_;
+  const rocksdb::Slice start_key_;
+  const rocksdb::Slice end_key_;
+  const int64_t max_keys_;
+  const DBTimestamp timestamp_;
+  const rocksdb::Slice txn_id_;
+  const uint32_t txn_epoch_;
+  const DBTimestamp txn_max_timestamp_;
+  const bool consistent_;
+  const bool check_uncertainty_;
+  DBScanResults results_;
+  std::unique_ptr<rocksdb::WriteBatch> kvs_;
+  std::unique_ptr<rocksdb::WriteBatch> intents_;
+  std::string key_buf_;
+  std::string saved_buf_;
+  bool peeked_;
+  bool is_get_;
+  cockroach::storage::engine::enginepb::MVCCMetadata meta_;
+  // cur_raw_key_ holds either iter_rep_->key() or the saved value of
+  // iter_rep_->key() if we've peeked at the previous key (and peeked_
+  // is true).
+  rocksdb::Slice cur_raw_key_;
+  // cur_key_ is the decoded MVCC key, separated from the timestamp
+  // suffix.
+  rocksdb::Slice cur_key_;
+  // cur_value_ holds either iter_rep_->value() or the saved value of
+  // iter_rep_->value() if we've peeked at the previous key (and
+  // peeked_ is true).
+  rocksdb::Slice cur_value_;
+  // cur_timestamp_ is the timestamp for a decoded MVCC key.
+  DBTimestamp cur_timestamp_;
+  int iters_before_seek_;
+};
+
+typedef mvccScanner<false> mvccForwardScanner;
+typedef mvccScanner<true> mvccReverseScanner;
+
+} // namespace cockroach
+
+// local variables:
```

Comment:
What's this?

Reply:
It tells emacs to use C++ mode. I should probably remove them. @bdarnell do you know of a better way to select C++ mode for some `.h` files and C mode for others?